jumpscale/sals/process/__init__.py

```diff
@@ -84,595 +141,833 @@ def execute(
+            return target in given
+        else:
+            return target.strip().lower() == given.lower()
 
-    if match_predicate is None:
-        match_predicate = default_predicate
+    default_processes_source = psutil.process_iter(["name", "exe", "cmdline"])
 
-    if process_name is None:
-        raise j.exceptions.RuntimeError("process cannot be None")
-    if j.data.platform.is_unix():
-        pids = set()
-        for process in get_processes():
-            try:
-                pid = process.pid
-                if not isinstance(pid, int):
-                    continue
-                name = process.name()
-                if match_predicate(process_name, name):
-                    pids.add(pid)
-                elif match_predicate(process_name, process.exe()):
-                    pids.add(pid)
-                else:
-                    cmdline = process.cmdline()
-                    if cmdline and cmdline[0]:
-                        if match_predicate(process_name, cmdline[0]):
-                            pids.add(pid)
-            except (psutil.Error, FileNotFoundError):
+    match_predicate = match_predicate or default_predicate
+    p_source = _alt_source or default_processes_source
+
+    pids = []
+    for proc in p_source:
+        try:
+            if not include_zombie and proc.status() == psutil.STATUS_ZOMBIE:
+                j.logger.debug(f"ignoring : {proc.pid} zombie process.")
                 continue
-        return list(pids)
-    else:
-        raise j.exceptions.NotImplemented("getProcessPid is only implemented for unix")
+            if (
+                (match_predicate(process_name, proc.info["name"]))
+                or (proc.info["exe"] and match_predicate(process_name, os.path.basename(proc.info["exe"])))
+                or (
+                    proc.info["cmdline"]
+                    and match_predicate(
+                        process_name,
+                        proc.info["cmdline"] if full_cmd_line else os.path.basename(proc.info["cmdline"][0]),
+                    )
+                )
+            ):
+                pids.append(proc.pid)
+                j.logger.debug(f"process {len(pids) }found: {proc.info}")
+                # return early if no need to iterate over all running process
+                if limit and len(pids) == limit:
+                    return pids
+        except (psutil.ZombieProcess, psutil.AccessDenied, psutil.NoSuchProcess):
+            pass
+    return pids
 
 
 def get_my_process():
-    """get process object of current process
+    """Get psutil.Process object of the current process.
 
     Returns:
-        [psutil.Process] -- process object
+        (psutil.Process): psutil.Process object of the current process.
     """
     return get_process_object(os.getpid())
 
 
 def get_process_object(pid, die=True):
-    """Get Process object of a process id
+    """Get psutil.Process object of a given process ID (PID).
```

Comment:
`die` option can be removed, either we raise an exception and mention this in the docstring or we return `None` (and I prefer the earlier).

Reply:
die option is the only purpose of this function. and it used when you want to initiate a Process and handle the exception if PID does not exist. instead of warp every calling to psuitl.Process() in try/except and handle the except, we call this function and check the bool return.

so instead of this
```
try:
    parent = psutil.Process(parent)
except (psutil.AccessDenied, psutil.NoSuchProcess):
    return
```

would be simplified
```
parent = get_process_object(parent, die=False)
if not parent:
    return
```

even simpler in 3.9
```
if parent:= get_process_object(parent, die=False):
    # do some thing with it
else:
    return
```

the only issue now, it needs to used properly in the code. let me work on it.

Reply:
code updated, now `get_process_object` used properly, and its `die` arg is needed in `get_processes_info()`.